### PR TITLE
Add character type mutations

### PR DESCRIPTION
**Pull request by backus**
Mutations:

  - `/\b/` → `/\B/`
  - `/\B/` → `/\b/`
  - `/\d/` → `/\D/`
  - `/\D/` → `/\d/`
  - `/\s/` → `/\S/`
  - `/\S/` → `/\s/`
  - `/\w/` → `/\W/`
  - `/\W/` → `/\w/`

### CODE REVIEW

**Comment by backus** [comment]
@mbj can you review?

**Comment by backus** [comment]
@mbj or @dkubb can you review?

File: lib/mutant/mutator/node/regexp/character_type.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+module Mutant
+  class Mutator
+    class Node
+      module Regexp
+        # Character type mutator
+        class CharacterType < Node
+          extend Mutant::AST::Sexp
```
Comment (mbj) [inline lib/mutant/mutator/node/regexp/character_type.rb]: This makes the AST DSL a public interface of this specific class. I wounder if we can do better.

Comment (backus) [inline lib/mutant/mutator/node/regexp/character_type.rb]: How? The module defines the methods as private so they aren't public interface: 

```ruby
2.3.0 :004 > Mutant::Mutator::Node::Regexp::CharacterType.s(:blah)
NoMethodError: private method `s' called for Mutant::Mutator::Node::Regexp::CharacterType:Class
	from (irb):4
	from /Users/johnbackus/.rvm/rubies/ruby-2.3.0/bin/irb:11:in `<main>'
2.3.0 :005 > Mutant::Mutator::Node::Regexp::CharacterType.n_not(:blah)
NoMethodError: private method `n_not' called for Mutant::Mutator::Node::Regexp::CharacterType:Class
```

Comment (mbj) [inline lib/mutant/mutator/node/regexp/character_type.rb]: I assumed it did cause a public interface. Thx for checking.

File: lib/mutant/mutator/node/regexp/character_type.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+          map = {
+            s(:regexp_word_type) => s(:regexp_nonword_type),
+            s(:regexp_digit_type) => s(:regexp_nondigit_type),
+            s(:regexp_space_type) => s(:regexp_nonspace_type),
+            s(:regexp_word_boundary_anchor) => s(:regexp_nonword_boundary_anchor)
+          }
```
Comment (mbj) [inline lib/mutant/mutator/node/regexp/character_type.rb]: What if we only map type to types and apply `s` in a map? This would reduce literal `s` method calls.

**Comment by backus** [comment]
@mbj ready for another pass

File: lib/mutant/mutator/node/regexp/character_type.rb (hunk 2 above)
Comment (backus) [inline lib/mutant/mutator/node/regexp/character_type.rb]: Done

**Comment by backus** [comment]
@mbj this is ready for review

File: meta/regexp/character_types.rb [hunk 3]
```diff
@@ -0,0 +1,111 @@
+Mutant::Meta::Example.add :regexp_word_type do
```
Comment (mbj) [inline meta/regexp/character_types.rb]: This is very repetitive, should we instead do something like this:

```
[ 
  :regexp_word_type, '\w', '\W',
  ....
].each do |node_type, original, replacement|
  Mutant::Meta::Example.add(node_type) do
    source("/#{original}/")
    singleton_mutations
    mutation('//')
    mutation '/nomatch\A'
    mutation("/#{replacement}/")
end
```

Its far easier to review this way.

File: lib/mutant/ast/regexp/transformer/direct.rb [hunk 4]
```diff
@@ -70,6 +70,7 @@ class ASTToExpression < Transformer::ASTToExpression
               [:regexp_script_hiragana_property, [:property, :script_hiragana,  '\p{Hiragana}'], ::Regexp::Expression::UnicodeProperty::Script],
               [:regexp_script_katakana_property, [:property, :script_katakana,  '\p{Katakana}'], ::Regexp::Expression::UnicodeProperty::Script],
               [:regexp_letter_any_property,      [:property, :letter_any,       '\p{L}'],        ::Regexp::Expression::UnicodeProperty::Letter::Any],
+              [:regexp_hex_type,                 [:type,     :hex,              '\h'],           ::Regexp::Expression::CharacterType::Hex],
```
Comment (mbj) [inline lib/mutant/ast/regexp/transformer/direct.rb]: Should we break this one up in a dedicated commit we can merge earlier?

**Comment by mbj** [comment]
@backus I left one comment for a possible reduction of spec duplication, and a possible PR extraction. Otherwise I'm fine with this change.

File: lib/mutant/ast/regexp/transformer/direct.rb (hunk 4 above)
Comment (backus) [inline lib/mutant/ast/regexp/transformer/direct.rb]: Sure

Comment (backus) [inline lib/mutant/ast/regexp/transformer/direct.rb]: Extracted to https://github.com/mbj/mutant/pull/634

File: meta/regexp/character_types.rb (hunk 3 above)
Comment (backus) [inline meta/regexp/character_types.rb]: Duplication resolved

**Comment by backus** [comment]
@mbj ready for another pass. Dependent on #635 being merged first

**Comment by backus** [comment]
@mbj duplication resolved and rebased. Dependent PR is merged so this is ready for review
